Clean up the JSObjectRef API

j-f1 opened the pull request:
Fixes #27.

- `get`/`set` → `dynamicMember` and subscripts
- `instanceof` (kept)
- `jsValue` (kept)
- `apply` → `callAsFunction`

j-f1 commented:
Inspired by the `ExpressibleByNilLiteral` protocol, I came up with this subscript-based `JSValueConvertible` protocol:

```swift
public protocol JSValueConvertible {
    subscript(jsValue _: ()) -> JSValue { get }
}

extension JSValue {
    public init(from convertible: JSValueConvertible) {
        self = convertible[jsValue: ()]
    }
}
```

j-f1 commented:
Alright, the tests are still failing because it seems like I can’t have `Array` conform to `JSValueConvertible` when either `Element: JSValueConvertible` or `Element == JSValueConvertible` :(

Any suggestions?

j-f1 commented:
Update: I’ve replaced all the function call/new logic with this pattern:

```swift
// JS version in comment
func() // func.apply(null, [])
func(arg1, arg2, arg3) // func.apply(null, [arg1, arg2, arg3])
func(args: [arg1, arg2, arg3]) // func.apply(null, [arg1, arg2, arg3])
func(this: foo) // func.apply(foo, [])
func(this: foo, arg1, arg2, arg3) // func.apply(foo, [arg1, arg2, arg3])
func(this: foo, args: [arg1, arg2, arg3]) // func.apply(foo,  [arg1, arg2, arg3])
func(new: arg1, arg2, arg3) // new func(arg1, arg2, arg3)
func(.new, args: [arg1, arg2, arg3]) // new func(arg1, arg2, arg3)
func(.new) // new func()
```

The last two I’m a little uncertain of.

j-f1 commented:
One last change I was going to make was changing the `instanceof` function to an operator. I tried it out with `<<` which isn’t super clear.

It seems Swift doesn’t allow letters in operators, so `instanceof` is out. There are a ton of Unicode characters allowed but the only ASCII ones are `!%&*+-/<=>?^|~`. Maybe `<<?` or `<==`?

kateinoigakukun commented:
Sorry, I didn't take a look much now, but could you ignore .swiftpm directory?

kateinoigakukun commented:
I agree with deprecating get and set and provide subscript. 

But on second thought, `new` and `instanceof` are not natural and not swifty interface.

This library provides JavaScript like APIs but also they should be swifty.

I think `instanceof`, `new` and `jsValue` are not common naming as members and if users want to use those names, they can use subscript accessor instead.

I disagree that APIs become unnatural caring about edge cases.

j-f1 commented:
Good point! I reverted those changes, so now the last few samples look like this:

```js
foo.new(arg1, arg2, arg3)
new foo(arg1, arg2, arg3)

foo.new(args: [arg1, arg2, arg3])
new foo(...[arg1, arg2, arg3])

foo.new()
new foo()
```

Additionally, I changed the JSValue code back to using `.jsValue()` and changed `instanceof` to add the method `isInstanceOf`.

kateinoigakukun commented on IntegrationTests/TestSuites/Sources/PrimaryTests/main.swift:
```diff
@@ -296,9 +296,9 @@ Object_Conversion: do {
+//    _ = jsArray2.push!(jsArray1)
 
-    try expectEqual(jsArray2[4], .object(jsArray1))
+//    try expectEqual(jsArray2[4], .object(jsArray1))
```
Could you revert these comments out?

kateinoigakukun commented on Sources/JavaScriptKit/JSFunction.swift:
```diff
@@ -1,45 +1,44 @@
     @discardableResult
-    public func dynamicallyCall(withArguments arguments: [JSValueConvertible]) -> JSValue {
-        let result = arguments.withRawJSValues { rawValues -> RawJSValue in
+    func callAsFunction(this: JSObjectRef? = nil, args: [JSValueConvertible]) -> JSValue {
```
```suggestion
    public func callAsFunction(this: JSObjectRef? = nil, args: [JSValueConvertible]) -> JSValue {
```

kateinoigakukun commented on Sources/JavaScriptKit/JSFunction.swift:
```diff
@@ -1,45 +1,44 @@
+    // a) the constructor explicitly returns an object, or
+    // b) the constructor returns nothing, which causes JS to return the `this` value, or
+    // c) the constructor returns undefined, null or a non-object, in which case JS also returns `this`.
+    public func new(args: [JSValueConvertible]) -> JSObjectRef {
```
```suggestion
    public func new(arguments: [JSValueConvertible]) -> JSObjectRef {
```

I think there shouldn't be abbreviations in public interface. 🙏
